File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/resources/files.ts
// @ts-nocheck
import type { FigmaClient } from '../figma-client.js';
import type { FigmaFile, FigmaFileRequestOptions } from './types.js';

function stringifyParams(
  opts: FigmaFileRequestOptions = {},
): Record<string, string | number | boolean> {
  const params: Record<string, string | number | boolean> = {};
  if (opts.version !== undefined) params.version = opts.version;
  if (opts.ids && opts.ids.length > 0) params.ids = opts.ids.join(',');
  if (opts.depth !== undefined) params.depth = opts.depth;
  if (opts.geometry !== undefined) params.geometry = opts.geometry;
  if (opts.plugin_data !== undefined) params.plugin_data = opts.plugin_data;
  if (opts.branch_data !== undefined) params.branch_data = opts.branch_data;
  return params;
}

export class FilesResource {
  constructor(private readonly client: FigmaClient) {}

  /** GET /v1/files/:key — full file document. */
  async get(fileKey: string, options: FigmaFileRequestOptions = {}): Promise<FigmaFile> {
    return this.client.get<FigmaFile>(`/v1/files/${encodeURIComponent(fileKey)}`, {
      params: stringifyParams(options),
    });
  }

  /** GET /v1/files/:key/nodes — subset of nodes. */
  async nodes(fileKey: string, ids: string[]): Promise<unknown> {
    return this.client.get<unknown>(`/v1/files/${encodeURIComponent(fileKey)}/nodes`, {
      params: { ids: ids.join(',') },
    });
  }

  /** GET /v1/images/:key — exported image URLs for selected nodes. */
  async images(
    fileKey: string,
    ids: string[],
    params: { format?: 'jpg' | 'png' | 'svg' | 'pdf'; scale?: number } = {},
  ): Promise<unknown> {
    const queryParams: Record<string, string | number> = { ids: ids.join(',') };
    if (params.format !== undefined) queryParams.format = params.format;
    if (params.scale !== undefined) queryParams.scale = params.scale;
    return this.client.get<unknown>(`/v1/images/${encodeURIComponent(fileKey)}`, {
      params: queryParams,
    });
  }

  /** GET /v1/files/:key/images — image fills uploaded to the file. */
  async imageFills(fileKey: string): Promise<unknown> {
    return this.client.get<unknown>(`/v1/files/${encodeURIComponent(fileKey)}/images`);
  }

  /** GET /v1/files/:key/meta — file metadata (name, version, thumbnail). */
  async meta(fileKey: string): Promise<unknown> {
    return this.client.get<unknown>(`/v1/files/${encodeURIComponent(fileKey)}/meta`);
  }

  /** GET /v1/files/:key/versions — version history (paginated). */
  async versions(
    fileKey: string,
    params: { page_size?: number; before?: string; after?: string } = {},
  ): Promise<unknown> {
    const queryParams: Record<string, string | number> = {};
    if (params.page_size !== undefined) queryParams.page_size = params.page_size;
    if (params.before !== undefined) queryParams.before = params.before;
    if (params.after !== undefined) queryParams.after = params.after;
    return this.client.get<unknown>(`/v1/files/${encodeURIComponent(fileKey)}/versions`, {
      params: queryParams,
    });
  }
}

File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/resources/files.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

if TYPE_CHECKING:
    from .._figma_client import FigmaClient


def _stringify_params(options: dict[str, Any] | None) -> dict[str, Any]:
    options = options or {}
    params: dict[str, Any] = {}
    if options.get("version") is not None:
        params["version"] = options["version"]
    if options.get("ids"):
        params["ids"] = ",".join(options["ids"])
    if options.get("depth") is not None:
        params["depth"] = options["depth"]
    if options.get("geometry") is not None:
        params["geometry"] = options["geometry"]
    if options.get("plugin_data") is not None:
        params["plugin_data"] = options["plugin_data"]
    if options.get("branch_data") is not None:
        params["branch_data"] = options["branch_data"]
    return params


class FilesResource:
    def __init__(self, client: FigmaClient) -> None:
        self._client = client

    async def get(self, file_key: str, **options: Any) -> dict[str, Any]:
        """GET /v1/files/:key — full file document."""
        return await self._client.get(
            f"/v1/files/{quote(file_key, safe='')}",
            params=_stringify_params(options),
        )

    async def nodes(self, file_key: str, ids: list[str]) -> dict[str, Any]:
        """GET /v1/files/:key/nodes — subset of nodes."""
        return await self._client.get(
            f"/v1/files/{quote(file_key, safe='')}/nodes",
            params={"ids": ",".join(ids)},
        )

    async def images(
        self,
        file_key: str,
        ids: list[str],
        *,
        format: str | None = None,
        scale: float | None = None,
    ) -> dict[str, Any]:
        """GET /v1/images/:key — export image URLs for selected nodes."""
        params: dict[str, Any] = {"ids": ",".join(ids)}
        if format is not None:
            params["format"] = format
        if scale is not None:
            params["scale"] = scale
        return await self._client.get(
            f"/v1/images/{quote(file_key, safe='')}",
            params=params,
        )

    async def image_fills(self, file_key: str) -> dict[str, Any]:
        """GET /v1/files/:key/images — image fills uploaded to the file."""
        return await self._client.get(f"/v1/files/{quote(file_key, safe='')}/images")

    async def meta(self, file_key: str) -> dict[str, Any]:
        """GET /v1/files/:key/meta — file metadata."""
        return await self._client.get(f"/v1/files/{quote(file_key, safe='')}/meta")

    async def versions(
        self,
        file_key: str,
        *,
        page_size: int | None = None,
        before: str | None = None,
        after: str | None = None,
    ) -> dict[str, Any]:
        """GET /v1/files/:key/versions — version history (paginated)."""
        params: dict[str, Any] = {}
        if page_size is not None:
            params["page_size"] = page_size
        if before is not None:
            params["before"] = before
        if after is not None:
            params["after"] = after
        return await self._client.get(
            f"/v1/files/{quote(file_key, safe='')}/versions", params=params
        )
